buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java

```diff
@@ -56,15 +57,18 @@
     private final Set<ElasticsearchCluster> runningClusters = new HashSet<>();
     private final Boolean allowClusterToSurvive = Boolean.valueOf(System.getProperty(TESTCLUSTERS_INSPECT_FAILURE, "false"));
 
+    private ReaperPlugin reaper;
+
     @Override
     public void apply(Project project) {
         project.getPlugins().apply(DistributionDownloadPlugin.class);
 
+        project.getRootProject().getPluginManager().apply(ReaperPlugin.class);
+        reaper = project.getRootProject().getPlugins().getPlugin(ReaperPlugin.class);
```

Comment:
Grabbing the plugin instance itself seems quite wrong. Can we put the bits we need here in a project extension applied by the plugin and use that instead?

Reply:
I was trying to keep the logic in the plugin, as was suggested previously. :) There isn't any actual configuration, which is why I didn't add an extension. I think we would want the state to remain in the plugin (the process instantiation) so what would the extension be other than a wrapper around its own ref to the plugin?

Reply:
Yeah, there's just something about passing around plugin instances that seems nasty to me. In fact, I wouldn't be surprised if these APIs are deprecated in the future. The `getPlugin()` method here is really meant for getting at the plugin convention objects (since replaced by extensions). What we really want is a singleton "service" but since Gradle's DI stuff isn't public we go with a workaround here. I think using an extension for the purpose is the best next option rather than passing references to a plugin instance. We are battling between the ideas the "plugins should be smart, extensions should be dumb" and "only Gradle should handle plugin instances". In the end I think the latter takes precedence over the former as I think these "service objects" are best implemented as extensions and not plugins.